diesel/src/connection/statement_cache/strategy.rs

```diff
@@ -0,0 +1,392 @@
+    fn default() -> Self {
+        Self {
+            cache: Default::default(),
+        }
+    }
+}
+
+impl<DB, Statement> StatementCacheStrategy<DB, Statement> for WithCacheStrategy<DB, Statement>
+where
+    DB: Backend,
+    StatementCacheKey<DB>: Hash + Eq,
+    DB::TypeMetadata: Clone,
+    DB::QueryBuilder: Default,
+{
+    fn get(
+        &mut self,
+        key: StatementCacheKey<DB>,
+        backend: &DB,
+        source: &dyn QueryFragmentForCachedStatement<DB>,
+        prepare_fn: &mut dyn FnMut(&str, bool) -> Result<Statement, Error>,
+    ) -> Result<MaybeCached<'_, Statement>, Error> {
+        let entry = self.cache.entry(key);
+        match entry {
+            Entry::Occupied(e) => Ok(MaybeCached::Cached(e.into_mut())),
+            Entry::Vacant(e) => {
+                let sql = e.key().sql(source, backend)?;
+                let st = prepare_fn(&sql, true)?;
+                Ok(MaybeCached::Cached(e.insert(st)))
+            }
+        }
+    }
+
+    fn cache_size(&self) -> CacheSize {
+        CacheSize::Unbounded
+    }
+}
+
+/// No statements will be cached,
+#[allow(missing_debug_implementations, unreachable_pub)]
+#[derive(Clone, Copy, Default)]
+pub struct WithoutCacheStrategy {}
+
+impl<DB, Statement> StatementCacheStrategy<DB, Statement> for WithoutCacheStrategy
+where
+    DB: Backend,
+    StatementCacheKey<DB>: Hash + Eq,
+    DB::TypeMetadata: Clone,
+    DB::QueryBuilder: Default,
+{
+    fn get(
+        &mut self,
+        key: StatementCacheKey<DB>,
+        backend: &DB,
+        source: &dyn QueryFragmentForCachedStatement<DB>,
+        prepare_fn: &mut dyn FnMut(&str, bool) -> Result<Statement, Error>,
+    ) -> Result<MaybeCached<'_, Statement>, Error> {
+        let sql = key.sql(source, backend)?;
+        Ok(MaybeCached::CannotCache(prepare_fn(&sql, false)?))
+    }
+
+    fn cache_size(&self) -> CacheSize {
+        CacheSize::Disabled
+    }
+}
+
+#[allow(dead_code)]
+#[cfg(test)]
+mod testing_utils {
+
+    use crate::{
+        connection::{Instrumentation, InstrumentationEvent},
+        Connection,
+    };
+
+    #[derive(Default)]
+    pub struct RecordCacheEvents {
+        pub list: Vec<String>,
+    }
+
+    impl Instrumentation for RecordCacheEvents {
+        fn on_connection_event(&mut self, event: InstrumentationEvent<'_>) {
+            if let InstrumentationEvent::CacheQuery { sql } = event {
+                self.list.push(sql.to_owned());
+            }
+        }
+    }
+
+    pub fn count_cache_calls(conn: &mut impl Connection) -> usize {
+        conn.instrumentation()
+            .as_any()
+            .downcast_ref::<RecordCacheEvents>()
+            .unwrap()
+            .list
+            .len()
+    }
+}
+
+#[cfg(test)]
+#[cfg(feature = "postgres")]
+mod tests_pg {
+    use crate::{
+        dsl::sql,
+        expression::functions::define_sql_function,
+        insertable::Insertable,
+        macros::table,
+        pg::Pg,
+        sql_types::{Integer, VarChar},
```

Comment:
We don't have such nested imports in diesel otherwise. I would prefer to keep the style consistent.